sync/syncer.go

```diff
@@ -189,98 +207,188 @@ func (s *Syncer) getLayerFromNeighbors(currenSyncLayer types.LayerID) (*types.La
 	}
 
-	blocksArr, err := s.fetchFullBlocks(blockIds)
-	if err != nil {
-		s.Error("could not get layer blocks %v", currenSyncLayer, err)
+	blocksArr := s.FetchFullBlocks(blockIds)
+	if len(blocksArr) == 0 {
+		s.Error("could not any blocks  for layer  %v", currenSyncLayer)
 		return nil, err
 	}
 
 	return types.NewExistingLayer(types.LayerID(currenSyncLayer), blocksArr), nil
 }
 
-func (s *Syncer) fetchFullBlocks(blockIds []types.BlockID) ([]*types.Block, error) {
-	output := s.fetchWithFactory(BlockReqFactory(blockIds), s.Concurrency)
+func (s *Syncer) FetchFullBlocks(blockIds []types.BlockID) []*types.Block {
+	output := s.fetchWithFactory(NewBlockWorker(s, s.Concurrency, BlockReqFactory(), blockSliceToChan(blockIds)))
 	blocksArr := make([]*types.Block, 0, len(blockIds))
 	for out := range output {
 		block := out.(*types.Block)
+		if err := s.ValidateBlock(block); err != nil {
+			s.Error(fmt.Sprintf("failed derefrencing block data %v %v", block.ID(), err))
+			continue
+		}
 
-		eligible, err := s.BlockEligible(&block.BlockHeader)
+		//data availability
+		txs, atxs, err := s.DataAvailabilty(block)
 		if err != nil {
-			s.Warning(fmt.Sprintf("failed checking eligiblety %v", block.ID()), err)
+			s.Error(fmt.Sprintf("data availabilty failed for block %v", block.ID()))
 			continue
 		}
-		if !eligible {
-			s.Warning(fmt.Sprintf("block %v not eligible", block.ID()), err)
+
+		//validate blocks view
+		if valid := s.ValidateView(block); valid == false {
+			s.Error(fmt.Sprintf("block %v not syntacticly valid", block.ID()))
 			continue
 		}
 
-		if err := s.syncMissingContent(block); err != nil {
-			s.Error(fmt.Sprintf("failed derefrencing block data %v", block.ID()), err)
+		if err := s.AddBlockWithTxs(block, txs, atxs); err != nil {
+			s.Error(fmt.Sprintf("failed to add block %v to database %v", block.ID(), err))
 			continue
 		}
 
+		s.Info(fmt.Sprintf("added block with txs to database %v", block.ID()))
+
 		s.Info("added block to layer %v", block.Layer())
 		blocksArr = append(blocksArr, block)
+	}
+
+	return blocksArr
+}
 
+func (s *Syncer) ValidateBlock(blk *types.Block) error {
+	s.Info("validate block %v ", blk.ID())
+
+	//check block signature and is identity active
+	if active, err := s.IsIdentityActive(blk.MinerID.Key, blk.Layer()); err != nil || !active {
+		return errors.New(fmt.Sprintf("block %v identity activation check failed ", blk.ID()))
+	}
+
+	//associated atx data availability
+	if blk.ATXID != *types.EmptyAtxId {
+		if _, err := s.syncAtxs([]types.AtxId{blk.ATXID}); err != nil {
+			return errors.New(fmt.Sprintf("cannot fetch associated atx for block %v %v ", blk.ID(), err))
+		}
+	}
+
+	//block eligibility
+	if eligable, err := s.BlockEligible(&blk.BlockHeader); err != nil || !eligable {
+		return errors.New(fmt.Sprintf("block %v eligablety check failed ", blk.ID()))
+	}
+
+	return nil
+}
+
+func (s *Syncer) ValidateView(blk *types.Block) bool {
+	vq := NewValidationQueue(s.Log.WithName("validation queue"))
```

Comment:
IMO this should rely solely on block validity that was determined when blocks were received to avoid recursive validation calls till the end of the mesh

Reply:
the whole ticket im handling is recursive validation, it was decided that this is what we are going to do, now is not the time to challenge that